crates/agentgateway/src/mcp/router.rs

```
@@ -149,12 +153,37 @@ impl App {
 		// MCP context is added later
 		req.extensions_mut().insert(Arc::new(ctx));
 
-		// Check if authentication is required and JWT token is missing
-		if let Some(auth) = &authn
-			&& req.extensions().get::<Claims>().is_none()
-			&& !Self::is_well_known_endpoint(req.uri().path())
-		{
-			return Self::create_auth_required_response(&req, auth).into_response();
+		// skip well-known OAuth endpoints for authn
+		if !Self::is_well_known_endpoint(req.uri().path()) {
+			let has_claims = req.extensions().get::<Claims>().is_some();
+			let validator = authn.as_ref().and_then(|a| a.jwt_validator.as_ref());
+
+			match (authn.as_ref(), validator, has_claims) {
+				// if mcp authn is configured, has a validator, and has no claims yet, validate
+				(Some(auth), Some(validator), false) => {
+					if let Ok(TypedHeader(Authorization(bearer))) = req
+						.extract_parts::<TypedHeader<Authorization<Bearer>>>()
+						.await
+					{
+						match validator.validate_claims(bearer.token()) {
+							Ok(claims) => {
+								req.headers_mut().remove(http::header::AUTHORIZATION);
+								req.extensions_mut().insert(claims);
+							},
+							Err(_e) => {
+								return Self::create_auth_required_response(&req, auth).into_response();
+							},
+						}
+					}
+					// MCP authn validation happens in optional mode, so if no token is present, do nothing
+				},
+				// if mcp authn is configured, has no validator, and no claims, reject
+				(Some(auth), None, false) => {
+					return Self::create_auth_required_response(&req, auth).into_response();
+				},
+				// if no mcp authn is configured or JWT already validated (claims exist), do nothing
+				_ => {},
```

hmm... so in theory now I can have 3 JWT policies: Gateway Route and MCPBackend.

The first 2 do not have this "allow if previous already did it". Maybe it _should_ given its impossible for it to pass if you have multiple, given we strip the token. But maybe we need to be consistent across all 3?

I am a bit worried about silently ignoring the auth though....

---

That's definitely a fair point. 

I'm okay with changing the MCP authn behavior to reject in cases where the token was stripped earlier on if we think multiple JWT policy configuration is a complete anti-pattern.

---

Just made the update to follow the behavior of the first 2 for the time being. We can always come back and address the other cases as a follow up later if there is demand.